src/gui/torrentcontentmodel.cpp

```diff
@@ -505,34 +505,45 @@ void TorrentContentModel::setupModelData(const BitTorrent::AbstractFileStorage &
     qDebug("Torrent contains %d files", filesCount);
     m_filesIndex.reserve(filesCount);
 
-    TorrentContentModelFolder *currentParent;
     QHash<TorrentContentModelFolder *, QHash<QString, TorrentContentModelFolder *>> folderMap;
+    QVector<QString> lastParentPath;
+    TorrentContentModelFolder *lastParent = m_rootItem;
     // Iterate over files
     for (int i = 0; i < filesCount; ++i)
     {
-        currentParent = m_rootItem;
         const QString path = info.filePath(i).data();
 
         // Iterate of parts of the path to create necessary folders
         QList<QStringView> pathFolders = QStringView(path).split(u'/', Qt::SkipEmptyParts);
-        pathFolders.removeLast();
+        const QString fileName = pathFolders.takeLast().toString();
 
-        for (const QStringView pathPart : asConst(pathFolders))
+        if (!std::equal(lastParentPath.begin(), lastParentPath.end()
```

Comment:
It looks like it can be further improved. Look, when the previous path was `a/b/c`, and the current one does not match it, then you "pessimistically" start the search from the beginning. But the current path can be, for example, `a/b`, `a`, or even `a/b/c/d`, which have a common part with the previous one.
However, I do not insist that you try to improve it in this PR. Apparently, it should have benefits in its current form. You took measurements, didn't you?

Reply:
> It looks like it can be further improved. Look, when the previous path was a/b/c, and the current one does not match it, then you "pessimistically" start the search from the beginning.

yes, it can be improved further, maybe for another PR sometime.

> , it should have benefits in its current form. You took measurements, didn't yo

yes, but can't remember the numbers, it was 4 months ago.